src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,123 @@
+#include <variant>
+#include <vector>
+
+#include "src/common/logging/logging.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+
+
+namespace raksha::ir::auth_logic {
+
+static Principal ConstructPrincipal(
+    AuthLogicParser::PrincipalContext& context) {
+  return Principal(context.ID()->getText());
+}
+
+static datalog::Predicate ConstructPredicate(
+    AuthLogicParser::PredicateContext& context) {
+  std::vector<std::string> args = utils::MapIter<std::string>(
+      context.pred_arg(),
+      [](auto* statement) { return statement->ID()->getText(); });
+  datalog::Sign sign = context.NEG() == nullptr ? datalog::Sign::kPositive
+                                                : datalog::Sign::kNegated;
+  return datalog::Predicate(context.ID()->getText(), args, sign);
+}
+
+static BaseFact ConstructVerbphrase(
+    Principal left_principal, AuthLogicParser::VerbphraseContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::PredphraseContext*>(
+          &context)) {  // PredicateContext
+    return BaseFact(Attribute(
+        left_principal, ConstructPredicate(*CHECK_NOTNULL(ctx->predicate()))));
+  }
+  auto& ctx =
+      *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::ActsAsPhraseContext*>(
+          &context));  // ActsAsPhraseContext
+  return BaseFact(CanActAs(
+      left_principal, ConstructPrincipal(*CHECK_NOTNULL(ctx.principal()))));
+}
+
+static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::PrinFactContext*>(&context)) {
+    return ConstructVerbphrase(
+        ConstructPrincipal(*CHECK_NOTNULL(ctx->principal())),
+        *CHECK_NOTNULL(ctx->verbphrase()));
+  }
+  auto& ctx = *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::PredFactContext*>(
+      &context));  // predFactContext
+  return BaseFact(ConstructPredicate(*CHECK_NOTNULL(ctx.predicate())));
+}
+
+static Fact ConstructFact(AuthLogicParser::FactContext& context) {
+  if (auto* flat_fact_fact_ctx =
+          dynamic_cast<AuthLogicParser::FlatFactFactContext*>(
+              &context)) {  // FlatFactFactContext
+    return Fact(
+        ConstructFlatFact(*CHECK_NOTNULL(flat_fact_fact_ctx->flatFact())));
+  }
+  auto& ctx = *CHECK_NOTNULL(
+      dynamic_cast<AuthLogicParser::CanSayFactContext*>(&context));
+  return ConstructFact(*CHECK_NOTNULL(ctx.fact()));
+}
+
+static Query ConstructQuery(AuthLogicParser::QueryContext& context) {
+  return Query(context.ID()->getText(),
+               ConstructPrincipal(*CHECK_NOTNULL(context.principal())),
+               ConstructFact(*CHECK_NOTNULL(context.fact())));
+}
+
+static Program ConstructProgram(AuthLogicParser::ProgramContext& context) {
+  std::vector<SaysAssertion> says_assertions;
```

Comment:
I guess parsing `says_assertions` is still WIP? It looks like these don't get built yet. (I just wanted to make sure this wasn't missed accidentally).

Reply:
Andrew, says_assertion is still a WIP and will create another PR after this loop is closed